src/utils/utils_concurrency.h

```diff
@@ -92,57 +101,123 @@ static __inline unsigned char utils_mssb_index(long long value) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    InterlockedExchangePointer(ptr, val);
+}
 
-#define utils_atomic_increment(object)                                         \
-    InterlockedIncrement64((LONG64 volatile *)object)
+static __inline uint64_t utils_atomic_increment_u64(uint64_t *ptr) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    // return incremented value
+    return InterlockedIncrement64((LONG64 volatile *)ptr);
+}
 
-#define utils_atomic_decrement(object)                                         \
-    InterlockedDecrement64((LONG64 volatile *)object)
+static __inline uint64_t utils_atomic_decrement_u64(uint64_t *ptr) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    // return decremented value
+    return InterlockedDecrement64((LONG64 volatile *)ptr);
+}
 
-#define utils_fetch_and_add64(ptr, value)                                      \
-    InterlockedExchangeAdd64((LONG64 *)(ptr), value)
+static __inline uint64_t utils_fetch_and_add_u64(uint64_t *ptr, uint64_t val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    // return the value that had previously been in *ptr
+    return InterlockedExchangeAdd64((LONG64 volatile *)(ptr), val);
+}
 
-// NOTE: windows version have different order of args
-#define utils_compare_exchange(object, desired, expected)                      \
-    InterlockedCompareExchange64((LONG64 volatile *)object, *expected, *desired)
+static __inline uint64_t utils_fetch_and_sub_u64(uint64_t *ptr, uint64_t val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    // return the value that had previously been in *ptr
+    // NOTE: on Windows there is no *Sub* version of InterlockedExchange
+    return InterlockedExchangeAdd64((LONG64 volatile *)(ptr), -(LONG64)val);
+}
+
+static __inline bool utils_compare_exchange_u64(uint64_t *ptr,
+                                                uint64_t *expected,
+                                                uint64_t *desired) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    LONG64 out = InterlockedCompareExchange64(
+        (LONG64 volatile *)ptr, *(LONG64 *)desired, *(LONG64 *)expected);
+    if (out == *(LONG64 *)expected) {
+        return true;
+    }
+
+    // else
+    *expected = out;
+    return false;
+}
 
 #else // !defined(_WIN32)
 
 #define utils_lssb_index(x) ((unsigned char)__builtin_ctzll(x))
 #define utils_mssb_index(x) ((unsigned char)(63 - __builtin_clzll(x)))
 
-#define utils_atomic_load_acquire(object, dest)                                \
-    do {                                                                       \
-        utils_annotate_acquire((void *)object);                                \
-        __atomic_load(object, dest, memory_order_acquire);                     \
-    } while (0)
+static inline void utils_atomic_load_acquire_u64(uint64_t *ptr, uint64_t *out) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)out, 8);
+    __atomic_load(ptr, out, memory_order_acquire);
```

Comment:
Why don't we call the `utils_annotate_acquire` anymore?

Reply:
removed by mistake - restored